### DISCUSSION TITLE
assets package - loading assets from relative paths

### DISCUSSION THREAD

**Discussion by Karuma303** [pixijs/pixijs#8642] [Q&A] [answered]
I have an issue with the latest pixi versions and the new assets package. I use pixi so do some generative art that I sell on fxhash.com. On this platform it is not allowed for your 'artwork' to load assets with absolute paths. You have to use relative paths otherwise your art project will not be accepted by the platform.

I had no issues with older pixi versions using the loader class but now with the new assets package my projects simply don't work anymore on fxhash.

So I did some inverstigation and found out that there is this code in the Loader class (/assets/src/loader/Loader.ts) where it loads the assets and in order to do that, any asset path is turned into an absolute path (line 124).
I found some other code in the assets framework where there was a comment like 'we cannot use relative paths in webworker' which might be the reason why all relative paths are turned into absolute paths.
So I made some research and found out, that using relative paths from webworkers is not really forbidden. The only thing is that relative paths are resolved from the path of the script where the webworker was started and not from the path where the script was loaded (typically the html page). For my special use case that would be no problem though, because the html page and scripts are based in the same directory.
I furthermore tried to fiddle around with basePath and rootPath of the asset resolver but without any success. I even had the feeling, that changing to rootPath did not have any effect at all.

So my question is now: How can I use the latest version of pixi to load assets via relative paths? Is that somehow possible in theory or do I have to stick to the old pixi version?

**Comment by bigtimebuddy** [comment]
Loading relative paths should be supported by assets. Can you provide an example of what you're trying to do and what's not working? Assets is still new and we are still ironing out the bugs.

**Reply by Karuma303** [reply]
There is no doubt that relative paths are supported. But as far as I can see they are turned into absolute paths before they are loaded:
`const url = utils.path.toAbsolute(asset.src);` (from packages/assets/src/loader/Loader.ts)

So here is a zip with a simple html page that uses pixi to load an image. 

[project.zip](https://github.com/pixijs/pixijs/files/9543669/project.zip)

This page runs perfectly fine if you run it from a local webserver. However, when I try to test it in the sandbox of fxhash, it does not work.
Give it a try, there's no subscription or something needed. Just open the [sandbox](https://www.fxhash.xyz/sandbox), drop the zip folder there and hit the 'start tests' button. You will see that the image is listed in the files window, but the artwork is not displayed. The browser dev tools will show you a 404 response error for the image.
You will see in the console output, that I added the relative path to the image.

My code is pretty straightforward:

```typescript
        console.log('path added to assets: ' + imgUrl)

        Assets.add('Background', imgUrl)

        await Assets.load(['Background'])

        const backgroundTexture = Assets.get(imgUrl)
```

As I said, I never had a problem with older versions of pixi when using the old loader. And fxhash.com just says "all paths must be relative" (see below)

![image](https://user-images.githubusercontent.com/24876752/189551414-5423124a-9c0d-47ca-b4cc-c91b4c4c6ab3.png)

In my previous message I mentioned that I found a comment somewhere, that Web Worker needs absolute paths. It's in the header of the utils class that the asset loader calls:

`packages/utils/src/path.ts`

```typescript
    /**
     * Converts URL to an absolute path.
     * When loading from a Web Worker, we must use absolute paths.
     * If the URL is already absolute we return it as is
     * If it's not, we convert it
     * @param url - The URL to test
     * @param customBaseUrl - The base URL to use
     * @param customRootUrl - The root URL to use
     */
    toAbsolute(url: string, customBaseUrl?: string, customRootUrl?: string)
    {
```

I hope that I really found an issue here and that my problems are not only based on my stupidity ;) I really like pixi and really appreciate all the work you guys put into it!

**Reply by Karuma303** [reply]
And maybe I should really setup a simple repo to demonstrate all of that...

**Reply by Karuma303** [reply]
And doing some further investigation I found out that with the old loader the referrer was always set in the HTTP request header. This is not the case anymore, when I load images with the new assets package. I wonder if this could be the source of my problems...

**Comment by Karuma303** [comment, marked as answer]
Update:
I solved my problem by cloning the `assets` package and by setting 
```typescript    
config: {
        preferWorkers: false,
,
```
inside of `loadTextures.ts`

So that means when testing inside of the sandbox of fxhash then no asset loading via worker is allowed. I search for differences in the requests when loading via worker and non-worker and the only thing I could find is that the worker does not send a referrer header. So maybe fxhash is expecting a referrer and blocks the request otherwise.
If I have some more time then I will  investigate this issue a bit further by adding a rerrer to fetch call of the worker (in `WorkerManager`)
